### PR TITLE
Stratified K-Fold Cross Validation

### LINKED ISSUE #31: Model validation

### PR DESCRIPTION
**Pull request by SeanLF**
Closes #31

### CODE REVIEW

File: active_learning/learn.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        y_train_with_missing = y_train + [None] * len(y_train_oracle)
+        # build dataset object for libact library
+        training_dataset = Dataset(X_train, y_train_with_missing)
+        oracle_training_dataset = Dataset(X_train, y_train + y_train_oracle.tolist())
+        # do active_learning and get trained model
+        trained_model = train_active_learning(num_queries, training_dataset, oracle_training_dataset)
+        # test model
+        predicted = trained_model.predict(X_test)
+        results.append({'y_true': y_test, 'y_pred': predicted})
+        print('Cross val iteration done\n')
```
Comment (SeanLF) [inline active_learning/learn.py]: To remove

File: active_learning/learn.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+from active_learning.libact_models import *
+
+
+def train_active_learning(num_queries, training_dataset, oracle_training_dataset):
+    active_learning_strategy = 1
+    query_strategy, strategy = get_active_learning_strategy(active_learning_strategy, training_dataset)
+    model = SGDClassifier(loss='hinge', penalty='l2', alpha=1e-2, n_iter=1000, random_state=None, learning_rate='optimal', class_weight='balanced')
+    # train on labeled instances
+    model.fit(*(training_dataset.format_sklearn()))
+    # active learning
```
Comment (markkasun) [inline active_learning/learn.py:37]: Try to reuse the code from learning.py. Create a separate method with condition statements that determine whether to produce intermediate results

File: active_learning/learn.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        training_dataset.update(query_id, label)
+        model.fit(*(training_dataset.format_sklearn()))
+    return model
+
+
+def cross_validate_active_learning(num_folds, X, y, labeled_instances, num_queries):
+    results = []
+    # make K folds
+    skf = StratifiedKFold(n_splits=num_folds)
+    skf.get_n_splits(X, y)
```
Comment (markkasun) [inline active_learning/learn.py]: Remove this line. Does nothing.

File: active_learning/learn.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    return results
+
+
+def preprocess(X_strings):
+    return numpy.array(TfidfVectorizer(stop_words='english').fit_transform(X_strings).toarray())
+
+
+def learn():
+    # Build dataset
+    dataset_name = 'Four University'
```
Comment (markkasun) [inline active_learning/learn.py]: Should be a parameter to be able to run on other datasets. Same with other variables below.

**Review by markkasun — CHANGES_REQUESTED** [review]

File: active_learning/learn.py [hunk 5]
```diff
@@ -79,7 +85,7 @@ def learn(**kwargs):
     dataset_id = kwargs.pop('dataset_id', DataSet.objects.get(name='Four University').id)
     num_queries = kwargs.pop('num_queries', 20)
     num_folds = kwargs.pop('num_folds', 10)
-    labeled_instances = kwargs.pop('dataset_id', 1)
+    labeled_instances = kwargs.pop('dataset_id', 5)
```
Comment (markkasun) [inline active_learning/learn.py]: 5?

File: active_learning/learning.py [hunk 6]
```diff
@@ -101,43 +101,28 @@ def learn(self, **kwargs):
         # Extract features from documents in the form of TF-IDF
         X = numpy.array(TfidfVectorizer(stop_words='english').fit_transform(self.X).toarray())
 
-        train_dataset, train_dataset_labeled, test_dataset, lookup_train, lookup_test, self.lookup_table = \
-            self.setup_datasets(X, self.y, self.lookup_table, train_size)
+        train_dataset, train_dataset_labeled, test_dataset, lookup_train, lookup_test, self.lookup_table = self.setup_datasets(X, self.y, self.lookup_table, train_size)
```
Comment (markkasun) [inline active_learning/learning.py:102]: Don't like the split lines? Much easier to read imo.

File: active_learning/custom_labelers.py [hunk 7]
```diff
@@ -62,3 +64,18 @@ def connect_handler(self, data):
...
         import json
         self.lbl = json.loads(data)['label']
+
+
+class AutoLabeler(Labeler):
+    """
+    Uses the ground truth to label instances
+    """
+
+    def __init__(self, **kwargs):
```
Comment (markkasun) [inline active_learning/custom_labelers.py]: Should make oracle_dataset a mandatory field?
     if self.oracle_dataset is None:
            raise TypeError(
                "__init__() missing required keyword-only argument: 'oracle_dataset'"
            )

File: active_learning/learn.py (hunk 5 above)
Comment (SeanLF) [reply active_learning/learn.py]: oh yeah, my bad

File: active_learning/learning.py (hunk 6 above)
Comment (SeanLF) [reply active_learning/learning.py:102]: nope, python multiline statements are gross 😛

File: active_learning/custom_labelers.py (hunk 7 above)
Comment (SeanLF) [reply active_learning/custom_labelers.py]: might have to move my params around

File: active_learning/learning.py (hunk 6 above)
Comment (markkasun) [reply active_learning/learning.py:102]: But think of the PEP8!

Comment (SeanLF) [reply active_learning/learning.py:102]: I increased the max line length... haha

**Review by markkasun — APPROVED** [review]

**Pull request merged**